### ISSUE TITLE
JavaScript text entry

### ISSUE THREAD

**Issue by titusfortner** [watir/watir#547]
Entering text via drivers tends to be slow. [(for reference)](http://stackoverflow.com/questions/25583641/set-value-of-input-instead-of-sendkeys-selenium-webdriver-nodejs)
Once we've verified a textfield or textarea can accept text, there isn't really a reason to type all of the text since that doesn't provide additional testing value.

I'm thinking perhaps `#set!` or `#set_text`, which would do the normal present/enabled/writeable checks, clear, perhaps type the first letter, use javascript to fill everything but the last character, then `send_keys` the final character. We can test to verify performance improvements for different drivers to see if it is worth doing.

This might solve the shadow dom issue people have had recently with React frameworks not recognizing text entry. I've also had issues in the past where the full text was not properly entered. Not sure if it is worth having an `#ensure_text` get added as well.

**Comment by azohra** [comment]
Holy Jeeze is JS ever fast. Just gave the code below a try across a form of 10 text fields. I see huge benefit to this if we mix in your ideas above to keep interaction somewhat 'real'. 

```ruby 
    def do(v)
        set v
    end

# VS

    def do!(v)
        browser.execute_script("arguments[0].value = '#{v}';", self)
    end
```

``` ruby 
  it 'do! better than do' do
    site.home_page.visit
    watir = Benchmark.measure { site.home_page.input.do(input_region_fields) }
    site.home_page.visit
    js = Benchmark.measure { site.home_page.visit.input.do!(input_region_fields) }
    expect(js.real).to be < watir.real
  end
```

In Chromedriver
```ruby 
       user     system      total        real
   0.080000   0.020000   0.100000 (  1.636390)  #watir
   0.030000   0.000000   0.030000 (  0.499667) #js-direct

**Comment by titusfortner** [comment]
is this chromedriver?

**Comment by azohra** [comment]
@titusfortner - yes apologies for missing that. I am expanding the test to encompass more element types, will keep adding data here.

**Comment by p0deje** [comment]
What is the problem with shadow DOM?

This sacrifices correctness (making sure all events like keyDown and keyUp are emitted) for performance, which is a bit too much IMO.

**Comment by titusfortner** [comment]
http://stackoverflow.com/questions/42774205/text-entered-with-watir-keeps-disappearing

But I see that as a side benefit more than a major feature.

The question here is what you are testing. Are you actually testing text entry from a keyboard standpoint, or data in a form submission? The answer could be different, but I wager the vast majority care much more about the latter than the former.

Then from a risk management standpoint, what is the likelihood that one keydown/keyup combo is going to work while others aren't? What is the real world concern where using this approach would fail to uncover an important bug?

Still, this shouldn't be the default behavior, but an option for people who know their forms aren't weird and just want text sent from browser to the backend.

**Comment by p0deje** [comment]
Okay, gotcha. I would propose then to have the following interface:

1. `#set` is kept what it is now.
2. `#set!` is doing the same with JS.

This way we could actually build the whole new bang interface for Watir.

**Comment by titusfortner** [comment]
Lol, what else do you want in the bang interface?

**Comment by p0deje** [comment]
Well, various things can be done using JavaScript:

1. `Element#click!` can emulate click event.
2. `Element#hover!` can emulate mouse over event.
3. `Cookie#set!` can change `document.cookie`.

Most interactions can be re-implemented using JavaScript. That is a shallow path though.

**Comment by titusfortner** [comment]
We should be able to do the first two with `#fire_event`
Not sure about cookie functionality.

**Comment by p0deje** [comment]
I know we can, but if I understand this topic correctly, the same can be done with `#execute_script`, so why bother creating `#set!`?

**Comment by titusfortner** [comment]
I don't want Watir users to have to know any xpath, let alone javascript. :-D

**Comment by p0deje** [comment]
Well then `#fire_event` should not be used as well because it often requires sending multiple events to simulate something and bang-methods can hide that from user.

**Comment by azohra** [comment]
`!` interface is exactly how we envisioned these changes. Danger ahead, proceed with caution.

**Comment by titusfortner** [comment]
Merged with #603
